core/vm/concrete_bindings_test.go

```diff
@@ -0,0 +1,211 @@
+	contract := NewContract(AccountRef(callerAddr), AccountRef(selfAddr), new(uint256.Int), 10_000_000)
+	ccVm := concreteEVM{evm: vm, contract: contract}
+
+	gas := uint64(10_000_000)
+	value := common.BytesToHash([]byte("value"))
+	// Bytecode that performs a storage opcode and return the value passed
+	statedb.SetCode(externalAddr, hexutil.MustDecode("0x6000356000355560006000355460005260206000F3"))
+	statedb.AddAddressToAccessList(externalAddr)
+
+	ret, remainingGas, err := ccVm.Call(externalAddr, value.Bytes(), gas, new(uint256.Int))
+	r.NoError(err)
+	r.Equal(ret, value.Bytes())
+	r.Less(remainingGas, gas)
+}
+
+func TestEVMCallDelegate(t *testing.T) {
+	var (
+		r            = require.New(t)
+		statedb, _   = state.New(types.EmptyRootHash, state.NewDatabase(rawdb.NewMemoryDatabase()), nil)
+		callerAddr   = common.BytesToAddress([]byte("caller"))
+		selfAddr     = common.BytesToAddress([]byte("self"))
+		externalAddr = common.BytesToAddress([]byte("external"))
+	)
+
+	vm := NewEVM(BlockContext{}, TxContext{}, statedb, params.TestChainConfig, Config{})
+	contract := NewContract(AccountRef(callerAddr), AccountRef(selfAddr), new(uint256.Int), 10_000_000)
+	ccVm := concreteEVM{evm: vm, contract: contract}
+
+	gas := uint64(10_000)
+	slot := common.BytesToHash([]byte("key"))
+	value := common.BytesToHash([]byte("value"))
+
+	// Bytecode to load the passed slot address from storage and return it
+	statedb.SetCode(externalAddr, hexutil.MustDecode("0x6000355460005260206000F3"))
+	statedb.SetState(selfAddr, slot, value)
+
+	ret, remainingGas, err := ccVm.CallDelegate(externalAddr, slot.Bytes(), gas)
+
+	r.NoError(err)
+	r.Equal(value.Bytes(), ret)
+	r.Less(remainingGas, gas)
+
+}
+
+func TestEVMCreate(t *testing.T) {
+	var (
+		r          = require.New(t)
+		statedb, _ = state.New(types.EmptyRootHash, state.NewDatabase(rawdb.NewMemoryDatabase()), nil)
+		callerAddr = common.BytesToAddress([]byte("caller"))
+		selfAddr   = common.BytesToAddress([]byte("self"))
+	)
+
+	vm := NewEVM(newBlockContext(), TxContext{}, statedb, params.TestChainConfig, Config{})
+	contract := NewContract(AccountRef(callerAddr), AccountRef(selfAddr), new(uint256.Int), 10_000_000)
+	ccVm := concreteEVM{
+		evm:      vm,
+		contract: contract,
+	}
+
+	gas := uint64(10_000)
+	code := hexutil.MustDecode("0x600260010160005260206000F3")
+
+	t.Run("Create", func(t *testing.T) {
+		ret, addr, remainingGas, err := ccVm.Create(code, gas, new(uint256.Int))
+
+		r.NoError(err, "Call should not return an error")
+		r.Equal(statedb.GetCode(addr), ret, "State not correct")
+		r.Less(remainingGas, gas, "Gas used should be less than the provided gas")
+	})
+
+	t.Run("Create2", func(t *testing.T) {
+		ret, addr, remainingGas, err := ccVm.Create2(code, gas, new(uint256.Int), uint256.NewInt(3521))
+
+		r.NoError(err, "Call should not return an error")
+		r.Equal(statedb.GetCode(addr), ret, "State not correct")
+		r.Less(remainingGas, gas, "Gas used should be less than the provided gas")
+	})
```

Comment:
- Expected value should go before actual value in r.Equal.
- The stored bytecode is checked against `ret` but `ret` isn't checked anywhere. If the binding was malfunctioning in a way that returns and empty byte slice and doesn't set the code, the test would still pass.
- We should check the address is correct (like you were doing before) since, even though we are not testing the EVM itself, we must check the binding is returning the correct value.
- There should be a comment describing what the bytecode does.
- Using a sum for the creation code is not wrong per se but it is not aligned with what a smart contract developer would expect when reading the code. Creation bytecode is expected to return runtime bytecode, not the result of a sum opperation.

We could, for example, do something like this:

```golang
gas := uint64(10_000)
creationCodePrefix := hexutil.MustDecode("0x600A600C600039600A6000F3") // Return the runtime code
runtimeCode := hexutil.MustDecode("0x607B60005260206000F3")            // Return 123
creationCode := append(creationCodePrefix, runtimeCode...)

t.Run("Create", func(t *testing.T) {
	ret, addr, remainingGas, err := ccEvm.Create(creationCode, gas, new(uint256.Int))
	r.NoError(err)
	r.Equal(runtimeCode, ret)
	r.Equal(runtimeCode, statedb.GetCode(addr))
	r.Equal(crypto.CreateAddress(selfAddr, 0), addr)
	r.Less(remainingGas, gas)
})
t.Run("Create2", func(t *testing.T) {
	salt := new(uint256.Int).SetBytes([]byte("salt"))
	ret, addr, remainingGas, err := ccEvm.Create2(creationCode, gas, new(uint256.Int), salt)
	r.NoError(err)
	r.Equal(runtimeCode, ret)
	r.Equal(runtimeCode, statedb.GetCode(addr))
	r.Equal(crypto.CreateAddress2(selfAddr, salt.Bytes32(), crypto.Keccak256(creationCode)), addr)
	r.Less(remainingGas, gas)
})
```